### ISSUE TITLE
Copper hangs when compiling a simple grammar

### ISSUE THREAD

**Issue by krame505** [melt-umn/copper#42]
See title, Copper didn't terminate when trying to compile the attached .copper file 
[Parser_edu_umn_cs_melt_metaocaml_driver_parse.copper.txt](https://github.com/melt-umn/copper/files/3082520/Parser_edu_umn_cs_melt_metaocaml_driver_parse.copper.txt)
(renamed so github lets me upload it).
This corresponds to the Silver grammar
```

terminal Identifier_t /[A-Za-z_\$][A-Za-z_0-9\$]*/;
terminal LParen_t '(';
terminal RParen_t ')';

terminal App_t '' association = left, precedence = 10;

nonterminal Expr_c with ast<Expr>, location;

concrete productions top::Expr_c
| e1::Expr_c e2::Expr_c
  operator=App_t
  { top.ast = appExpr(e1.ast, e2.ast, location=top.location); }
| id::Identifier_t
  { top.ast = varExpr(id.lexeme, location=top.location); }
```
Also, as a side question, does Copper support setting the associativity of a production lacking terminals using the `<Operator>` field?  It seems like it should but the behavior I am seeing seems to indicate not.

**Comment by krame505** [comment]
It also appears that this issue persists even removing the `<Operator>` field.

**Comment by schwerdf** [comment]
This grammar produces an LR DFA with a state that has a goto transition to itself -- something Copper has apparently never encountered before.

You should be able to set the associativity of a terminal-less production the same way as you set the precedence.

**Comment by krame505** [comment]
Interesting, thanks for taking a look.  

Associativity and precedence are both specified on a production using `<Operator>`, right?  Then why does the above grammar still yield a shift-reduce conflict when using the new version of Copper?

**Comment by schwerdf** [comment]
Because the shift-reduce conflict is on `Identifier_t` (the only used terminal in the grammar), and `Identifier_t` has no assigned precedence.

**Comment by krame505** [comment]
Ah, nevermind then.  Thanks.

**Comment by ericvanwyk** [comment]
But should Copper ever fail to terminate when compiling a grammar?  Even if
the grammar has problems Copper should not hang.

**Comment by schwerdf** [comment]
The non-terminating behavior was a bug (fixed in commit ddfe1dd) unrelated to the precedence issue.

**Comment by ericvanwyk** [comment]
OK, thanks.  I was reading this in its email form and that didn't include the mention about the commit.  But I see it only when I go to the web interface for this.  Thanks.
